src/Lean/Server/References.lean

```diff
@@ -0,0 +1,243 @@
+
+  refs.foldl (init := #[]) fun refs ref => match ref with
+    | { ident := RefIdent.fvar id, range, isDeclaration := true } => Id.run do
+      -- Since deduplication works via definitions, we know that a definition
+      -- for the base id exists.
+      if let some baseId := idMap.find? id then
+        if id == baseId then -- Only keep the base definition
+          return refs.push ref
+      return refs
+    | { ident := ident@(RefIdent.fvar _), range, isDeclaration := false } =>
+      refs.push { ident := applyIdMap idMap ident, range, isDeclaration := false }
+    | _ => refs.push ref
+  where
+    applyIdMap : HashMap FVarId FVarId → RefIdent → RefIdent
+      | m, RefIdent.fvar id => RefIdent.fvar <| m.findD id id
+      | _, ident => ident
+
+def findModuleRefs (text : FileMap) (trees : List InfoTree) (localVars : Bool := true)
+    : ModuleRefs := Id.run do
+  let mut refs := combineFvars <| findReferences text trees
+  if !localVars then
+    refs := refs.filter fun
+      | { ident := RefIdent.fvar _, .. } => false
+      | _ => true
+  refs.foldl (init := HashMap.empty) fun m ref => m.addRef ref
+
+/- Collecting and maintaining reference info from different sources -/
+
+structure References where
+  bundles : HashMap System.FilePath Bundle
+  /-- Replaces the corresponding modules in `bundles` -/
+  overlays : HashMap Name (Nat × ModuleRefs)
+
+namespace References
+
+def empty : References := { bundles := HashMap.empty, overlays := HashMap.empty }
+
+def addBundle (self : References) (path : System.FilePath) (bundle : IleanBundle) : References :=
+  { self with bundles := self.bundles.insert path bundle.toBundle }
+
+def removeBundle (self : References) (path : System.FilePath) : References :=
+  { self with bundles := self.bundles.erase path }
+
+def addOverlay (self : References) (module : Name) (version : Nat) (refs : ModuleRefs)
+    : References := Id.run do
+  if let some (oldVersion, _) := self.overlays.find? module then
+    if version <= oldVersion then
+      return self -- Trying to add an older version
+  return { self with overlays := self.overlays.insert module (version, refs) }
+
+def removeOverlay (self : References) (module : Name) : References :=
+  { self with overlays := self.overlays.erase module }
+
+def overlayBundle (self : References) : Bundle :=
+  self.overlays.toList.foldl (init := Bundle.empty) fun bundle (name, _, refs) => bundle.addModule name refs
+
+/-- All bundles and the overlay combined into a single bundle -/
+def bundle (self : References) : Bundle :=
+  let base := self.bundles.toList.foldl (init := Bundle.empty) fun l (_, r) => l.addBundle r
+  base.addBundle self.overlayBundle
+
+def findAt? (self : References) (module : Name) (pos : Lsp.Position) : Option RefIdent := Id.run do
+  if let some refs := self.bundle.find? module then
+    return refs.findAt? pos
+  none
+
+def referingTo (self : References) (ident : RefIdent) (srcSearchPath : SearchPath)
```

Comment:
Let's not repeat [HTTP's mistake](https://en.wikipedia.org/wiki/HTTP_referer#Etymology) :)